File: ConsoleApp/src/main/java/overturn/ELConfig.java
package overturn;

import org.apache.commons.io.IOUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.AnnotationConfigApplicationContext;
import org.springframework.context.annotation.ComponentScan;
import org.springframework.context.annotation.Configuration;
import org.springframework.context.annotation.PropertySource;
import org.springframework.context.support.PropertySourcesPlaceholderConfigurer;
import org.springframework.core.env.Environment;
import org.springframework.core.io.Resource;

import java.nio.charset.Charset;

@Configuration
@ComponentScan
@PropertySource("test.properties")
public class ELConfig {

    /**
     * 为了使用占位符， 我们必须要配置一个 PropertyPlaceholderConfigurer bean 或 PropertySourcesPlaceholderConfigurer bean
     * 从 Spring 3.1 开始， 推荐使用 PropertySourcesPlaceholderConfigurer， 因为它能够基于 Spring Environment 及其属性源来解析占位符
     */

    // 注入普通
    @Value("I Love You!")
    private String norma;

    // 注入操作系统属性
    @Value("#{systemProperties['os.name']}")
    private String osName;

    // 输入表达式运算结果
    @Value("#{T(java.lang.Math).random() * 100.0}")
    private double randomNumber;

    // 输入其他 Bean 的属性
    // Bean 名称默认为首字母小写，也可以自行指定
    @Value("#{demoService.another}")
    private String fromAnother;

    // 注入文件内容
    @Value("classpath:test.txt")
    private Resource testFile;

    // 注入网址内容
    @Value("http://www.baidu.com/")
    private Resource testUrl;

    // 注入属性内容
    @Value("${book.name}")
    private String bookName;

    // 注入环境变量
    @Autowired
    private Environment environment;

    public static PropertySourcesPlaceholderConfigurer propertySourcesPlaceholderConfigurer() {
        return new PropertySourcesPlaceholderConfigurer();
    }

    public static void main(String[] args) {
        AnnotationConfigApplicationContext context =
                new AnnotationConfigApplicationContext(ELConfig.class);
        // Config 默认就是一个 Bean
        ELConfig resourceService = context.getBean(ELConfig.class);
        resourceService.outputResource();
        context.close();
    }

    public void outputResource() {
        try {
            System.out.println(norma);
            System.out.println(osName);
            System.out.println(randomNumber);
            System.out.println(fromAnother);
            System.out.println(IOUtils.toString(testFile.getInputStream(), Charset.forName("utf-8")));
            System.out.println(IOUtils.toString(testUrl.getInputStream(), Charset.forName("utf-8")));
            System.out.println(bookName);
            System.out.println(environment.getProperty("book.author"));
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}

File: BasicConsoleApp/src/main/kotlin/basicconsoleapp/config/ELConfig.kt
package basicconsoleapp.config

import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.annotation.Value
import org.springframework.context.annotation.Configuration
import org.springframework.context.annotation.PropertySource
import org.springframework.core.env.Environment
import org.springframework.core.io.Resource

/**
 * Created by ZhangJin on 2018/6/2.
 */


// 里面的属性都是 EL 注入进去的
@Configuration
// 导入外部属性源
// 这样就能通过 ${book.name} 语法进行注入了
// 外部源中的值，同样也会填充到 Environment 中
@PropertySource("test.properties")
open class ELConfig {

    // 注入一个字符串常量
    @Value("I Love You!")
    lateinit var normal: String

    // 注入 systemProperties
    @Value("#{systemProperties['os.name']}")
    lateinit var osNameString: String

    // 生成一个随机数
    @Value("#{T(java.lang.Math).random() * 100.0}")
    var randomNumber: Double = 0.0

    // 注入一个 Bean 对象中属性
    // 此处直接使用了 Bean 对象的默认名称
    @Value("#{demoService.another}")
    lateinit var fromAnother: String

    // 包中的文件
    // 注入为 InputStream
    @Value("test.txt")
    lateinit var testFile: Resource

    // 远程 URL 访问
    // 注入为 InputStream
    @Value("http://www.baidu.com")
    lateinit var testUrl: Resource

    // 直接注入 test.properties 中的数据
    // 由于 Kotlin 语法的关系，此处需要对 $ 进行转义
    @Value("\${book.name}")
    lateinit var bookName: String

    // test.properties 中的数据
    // 也会注入到 Environment 中
    @Autowired
    lateinit var env: Environment

}